provider/datagen/src/source.rs

```diff
@@ -132,44 +150,93 @@ impl Debug for TomlCache {
 }
 
 impl TomlCache {
-    pub fn new(root: PathBuf) -> Self {
+    pub fn new(root: AbstractFs) -> Self {
         Self {
             root,
             cache: Arc::new(FrozenMap::new()),
         }
     }
 
-    pub fn read_and_parse_toml<S, P: AsRef<Path>>(&self, path: P) -> Result<&S, DataError>
+    pub fn read_and_parse_toml<S>(&self, path: &str) -> Result<&S, DataError>
     where
         for<'de> S: serde::Deserialize<'de> + 'static + Send + Sync,
     {
-        let path = self.root.join(path);
-
-        if self.cache.get(&path).is_none() {
-            log::trace!("Reading: {:?}", &path);
-            let file = std::fs::read_to_string(&path)
-                .map_err(|e| DataError::from(e).with_path_context(&path))?;
-            let file: S = toml::from_str(&file)
-                .map_err(|e| crate::error::data_error_from_toml(e).with_path_context(&path))?;
-            self.cache.insert(path.clone(), Box::new(file));
+        if self.cache.get(path).is_none() {
+            let file = self.root.open(path)?;
+            let file: S = toml::from_slice(&file)
+                .map_err(|e| crate::error::data_error_from_toml(e).with_display_context(&path))?;
+            self.cache.insert(path.to_string(), Box::new(file));
         }
         self.cache
-            .get(&path)
+            .get(path)
             .unwrap()
             .downcast_ref::<S>()
             .ok_or_else(|| DataError::custom("Uprops TOML error").with_type_context::<S>())
     }
 
     #[cfg_attr(not(feature = "experimental"), allow(dead_code))]
     pub fn list(&self) -> Result<impl Iterator<Item = PathBuf>, DataError> {
-        let mut result = vec![];
-        for entry in std::fs::read_dir(&self.root)
-            .map_err(|e| DataError::from(e).with_path_context(&self.root))?
-        {
-            let entry = entry.map_err(|e| DataError::from(e).with_path_context(&self.root))?;
-            let path = entry.path();
-            result.push(path);
+        self.root.list("").map(Vec::into_iter)
+    }
+}
+
+#[derive(Debug)]
+pub(crate) enum AbstractFs {
+    Fs(PathBuf),
+    // The PB points to a valid ZIP file, the String is a prefix inside the ZIP
+    Zip(PathBuf, String),
+}
+
+impl AbstractFs {
+    pub fn new<P: AsRef<Path>>(root: P, prefix: String) -> Result<Self, DataError> {
+        if std::fs::metadata(root.as_ref())?.is_dir() {
+            Ok(Self::Fs(root.as_ref().join(prefix)))
+        } else {
+            zip::ZipArchive::new(File::open(&root)?)
+                .map_err(|e| DataError::custom("Zip").with_display_context(&e))?;
+            Ok(Self::Zip(root.as_ref().into(), prefix))
         }
-        Ok(result.into_iter())
+    }
+
+    pub fn open(&self, path: &str) -> Result<Vec<u8>, DataError> {
```

Comment:
Suggestion: name this function `read_to_buf` or similar. It's misleading because `open` typically means you need to `close` it later, which you don't here.